infrastructure/src/main/java/org/corfudb/infrastructure/log/statetransfer/batchprocessor/StateTransferBatchProcessor.java

```diff
@@ -43,54 +48,68 @@
      */
     default TransferBatchResponse writeRecords(ReadBatch readBatch, LogUnitClient logUnitClient,
                                                int writeRetriesAllowed, Duration writeSleepDuration) {
-        List<Long> totalAddressesInRequest = readBatch.getAddresses();
-        Optional<String> destination = readBatch.getDestination();
+        if (readBatch == null || readBatch.getAddresses().isEmpty()) {
+            throw new IllegalArgumentException("The read batch is empty.");
+        }
+        List<Long> totalAddressesInRequest = ImmutableList.copyOf(readBatch.getAddresses());
         Optional<Exception> lastWriteException = Optional.empty();
         Optional<TransferBatchResponse> transferBatchResponse = Optional.empty();
 
         for (int i = 0; i < writeRetriesAllowed; i++) {
             List<LogData> remainingDataToWrite = readBatch.getData();
             try {
-                boolean writeSucceeded = logUnitClient.writeRange(remainingDataToWrite).join();
-                if (!writeSucceeded) {
-                    throw new IllegalStateException("Failed to write to a log unit server.");
-                } else {
-                    lastWriteException = Optional.empty();
+                CFUtils.getUninterruptibly(logUnitClient.writeRange(remainingDataToWrite),
+                        OverwriteException.class,
+                        TimeoutException.class,
+                        NetworkException.class);
+                transferBatchResponse = Optional.of(TransferBatchResponse
+                        .builder()
+                        .transferBatchRequest(new TransferBatchRequest(totalAddressesInRequest,
+                                readBatch.getDestination()))
+                        .status(SUCCEEDED)
+                        .build());
+                break;
+            } catch (OverwriteException | TimeoutException | NetworkException e) {
+                lastWriteException = Optional.of(e);
+                Sleep.sleepUninterruptibly(writeSleepDuration);
```

Comment:
It would be great if you bring CFUtils changes from your second StateTransfer PR.  
Then you can get rid of sleep's (at least here)

P.S. also those changes in CFUtils are very useful, I miss it often